File: c++/minimum_height_trees.cpp
/*
For a undirected graph with tree characteristics, we can choose any node as the root.
The result graph is then a rooted tree. Among all possible rooted trees, those with minimum 
height are called minimum height trees (MHTs). Given such a graph, write a function to find
all the MHTs and return a list of their root labels.

Format
The graph contains n nodes which are labeled from 0 to n - 1. You will be given the number n
and a list of undirected edges (each edge is a pair of labels).

You can assume that no duplicate edges will appear in edges. Since all edges are undirected,
[0, 1] is the same as [1, 0] and thus will not appear together in edges.

Example 1:

Given n = 4, edges = [[1, 0], [1, 2], [1, 3]]

        0
        |
        1
       / \
      2   3
return [1]

Example 2:

Given n = 6, edges = [[0, 3], [1, 3], [2, 3], [4, 3], [5, 4]]

     0  1  2
      \ | /
        3
        |
        4
        |
        5
return [3, 4]

Hint:

How many MHTs can a graph have at most?
Note:

(1) According to the definition of tree on Wikipedia: “a tree is an undirected graph in which any
two vertices are connected by exactly one path. In other words, any connected graph without simple
cycles is a tree.”

(2) The height of a rooted tree is the number of edges on the longest downward path between the
root and a leaf.
*/

#include <vector>
#include <iostream>
#include <queue>

using namespace std;

class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<pair<int, int>>& edges) { 
        if (n == 1) return {0};
        vector<int> inDegree; // inDegree of each node
        inDegree.resize(n, 0);

        vector<vector<int>> graph;
        graph.resize(n);

        // build a graph and populate inDegree for each node
        for (auto& p : edges) {
            graph[p.first].push_back(p.second);
            ++inDegree[p.second];
            graph[p.second].push_back(p.first);
            ++inDegree[p.first];
        }

        queue<int> leafQueue;
        for (int i = 0; i < inDegree.size(); ++i) {
            if (inDegree[i] == 1) {
                leafQueue.push(i);
            }
        }

        while (n > 2) {
            int sz = leafQueue.size();
            for (int i = 0; i < sz; ++i) {
                --n;
                int node = leafQueue.front();
                leafQueue.pop();
                inDegree[node] = 0;
                for (auto neighbor : graph[node]) {
                    if (inDegree[neighbor] > 0 && --inDegree[neighbor] == 1) {
                        leafQueue.push(neighbor);
                    }
                }
            }
        }

        vector<int> res;
        while (!leafQueue.empty()) {
            res.push_back(leafQueue.front());
            leafQueue.pop();
        }
        return res;
    }
};

int main()
{
    vector<pair<int,int>> edges;
    edges.push_back(make_pair(0, 1));
    edges.push_back(make_pair(0, 2));
    edges.push_back(make_pair(0, 3));
    edges.push_back(make_pair(3, 4));
    edges.push_back(make_pair(4, 5));
    int n = 6;
    Solution sol;
    vector<int> res = sol.findMinHeightTrees(n, edges);
    for (int v : res) {
        cout << v << ' ';
    }
    cout << endl;
    return 0;
}

File: java/MinimumHeightTrees.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Created on 8/7/16.
 * For a undirected graph with tree characteristics, we can choose any node as the root.
 * The result graph is then a rooted tree. Among all possible rooted trees, those with minimum
 * height are called minimum height trees (MHTs). Given such a graph, write a function to find
 * all the MHTs and return a list of their root labels.
 *
 * Format
 * The graph contains n nodes which are labeled from 0 to n - 1. You will be given the number n
 * and a list of undirected edges (each edge is a pair of labels).
 *
 * You can assume that no duplicate edges will appear in edges. Since all edges are undirected,
 * [0, 1] is the same as [1, 0] and thus will not appear together in edges.
 *
 * Example 1:
 *
 * Given n = 4, edges = [[1, 0], [1, 2], [1, 3]]
 *
 * 0
 * |
 * 1
 * / \
 * 2   3
 * return [1]
 *
 * Example 2:
 *
 * Given n = 6, edges = [[0, 3], [1, 3], [2, 3], [4, 3], [5, 4]]
 *
 * 0  1  2
 * \ | /
 * 3
 * |
 * 4
 * |
 * 5
 * return [3, 4]
 *
 * Hint:
 *
 * How many MHTs can a graph have at most?
 * Note:
 *
 * (1) According to the definition of tree on Wikipedia: “a tree is an undirected graph in which any
 * two vertices are connected by exactly one path. In other words, any connected graph without simple
 * cycles is a tree.”
 *
 * (2) The height of a rooted tree is the number of edges on the longest downward path between the
 * root and a leaf.
 */

public class MinimumHeightTrees {

    public static void main(String[] args) {
        int n = 4;
        int[][] edges = {{1, 0}, {1, 2}, {1, 3}};
        MinimumHeightTrees app = new MinimumHeightTrees();
        System.out.println(app.findMinHeightTrees(n, edges).toString());
    }

    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> res = new ArrayList<>();
        if (n == 1) {
            res.add(0);
            return res;
        }
        int[] degree = new int[n]; // in degree of each node
        List<List<Integer>> graph = new ArrayList<>(n);
        for (int i = 0; i < n; ++i) {
            graph.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            ++degree[edge[0]];
            graph.get(edge[0]).add(edge[1]);
            ++degree[edge[1]];
            graph.get(edge[1]).add(edge[0]);
        }

        Queue<Integer> leafQueue = new LinkedList<>();
        for (int i = 0; i < n; ++i) {
            if (degree[i] == 1) leafQueue.offer(i);
        }

        while (n > 2) {
            int sz = leafQueue.size();
            for (int i = 0; i < sz; ++i) {
                --n;
                int curNode = leafQueue.poll();
                --degree[curNode];
                List<Integer> neighbors = graph.get(curNode);
                for (Integer neighbor : neighbors) {
                    if (degree[neighbor] > 0 && --degree[neighbor] == 1) {
                        leafQueue.offer(neighbor);
                    }
                }
            }
        }

        while (!leafQueue.isEmpty()) {
            res.add(leafQueue.poll());
        }
        return res;
    }
}
